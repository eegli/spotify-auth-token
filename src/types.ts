type OptionalConfig = {
  port: number;
  scopes: string;
  outFileName: string;
  outDir: string;
  show: boolean;
};

type RequiredConfig = {
  clientId: string;
  clientSecret: string;
};

export type UserConfig = RequiredConfig & Partial<OptionalConfig>;
export type AppConfig = RequiredConfig & OptionalConfig;

export type SpotifyTokenResponse = {
  access_token: string;
  token_type: 'Bearer';
  expires_in: number;
  refresh_token: string;
  scope: string;
  date_obtained: string;
};
