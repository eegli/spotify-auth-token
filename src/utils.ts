import { existsSync, mkdirSync, writeFileSync } from 'fs';
import { join } from 'path';

export function objectFromEntries<
  Key extends PropertyKey,
  Entries extends Array<[Key, unknown]>
>(
  value: Entries
): {
  [K in Extract<Entries[number], [Key, unknown]>[0]]: Extract<
    Entries[number],
    [K, unknown]
  >[1];
} {
  return Object.fromEntries(value) as {
    [K in Extract<Entries[number], [Key, unknown]>[0]]: Extract<
      Entries[number],
      [K, unknown]
    >[1];
  };
}

export function write(
  relativePath: string,
  fileName: string,
  data: unknown
): string {
  let path = join(process.cwd(), relativePath, fileName);
  if (!existsSync(path)) {
    mkdirSync(path, { recursive: true });
  }
  if (!path.endsWith('.json')) {
    path += '.json';
  }

  writeFileSync(path, JSON.stringify(data, null, 2));
  return path;
}

export function id(): string {
  return Math.random().toString(36).slice(2);
}

export function goodBye(message: string): never {
  console.error('\x1b[31m', `Error: ${message}. Goodbye`, '\x1b[0m');
  process.exit(1);
}
