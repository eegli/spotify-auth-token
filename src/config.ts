import { AppConfig } from './types';
import { goodBye } from './utils';

export const defaultConfig: AppConfig = {
  clientId: '',
  clientSecret: '',
  port: 3000,
  outDir: '',
  outFileName: 'spotify-token',
  scopes: 'user-read-email',
  show: false,
};

export function configFactory<
  T extends Record<string, string | number | boolean>,
  R extends keyof T
>(base: T, ...required: Array<R extends string ? R : never>) {
  return function (args: Partial<T> | Array<string>): T {
    const cfmap = new Map(Object.entries(base));

    // [--id, 123, --secret, xyz] --> arg, argVal, arg, argVal

    if (!Array.isArray(args)) {
      Object.entries(args).forEach(([arg, argVal]) => {
        if (cfmap.has(arg)) {
          const configVal = cfmap.get(arg);
          if (typeof configVal === typeof argVal) {
            cfmap.set(arg, argVal);
          } else {
            goodBye(
              `Invalid type for option "${configVal}". Expected ${typeof configVal}, got ${typeof argVal}`
            );
          }
        } else {
          console.warn(`Ignoring unknown option "${arg}"`);
        }
      });
    } else {
      args.forEach((val, idx, orig) => {
        if (val.startsWith('--')) {
          const arg = val.slice(2);
          const key = cfmap.get(arg);
          if (cfmap.has(arg)) {
            let argVal: string | number | boolean = orig[idx + 1];
            // Boolean flag that default to false
            if (typeof key === 'boolean') {
              cfmap.set(arg, true);
            }
            // Convert string input to number
            else if (typeof key === 'number') {
              cfmap.set(arg, +argVal);
            } else if (typeof key === typeof argVal) {
              cfmap.set(arg, argVal);
            } else {
              goodBye(
                `Invalid type for option "--${arg}". Expected ${typeof key}, got ${typeof argVal}`
              );
            }
          } else {
            console.warn(`Ignoring unknown option "--${arg}"`);
          }
        }
      });
    }

    // Check for required properties
    required.forEach((prop) => {
      if (!cfmap.get(prop)) {
        goodBye(`Missing required config property "${prop}"`);
      }
    });

    return Object.fromEntries(cfmap) as T;
  };
}
