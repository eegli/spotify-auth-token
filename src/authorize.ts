import { configFactory, defaultConfig } from './config';
import { getLocalhostUrl, request } from './request';
import type { SpotifyTokenResponse, UserConfig } from './types';
import { goodBye, id, write } from './utils';

const createConfig = configFactory(defaultConfig, 'clientId', 'clientSecret');

export async function authorize(userConfig: UserConfig): Promise<void> {
  try {
    const config = userConfig
      ? createConfig(userConfig)
      : createConfig(process.argv.slice(2));

    const redirectUri = `http://localhost:${config.port}`;
    const state = id();

    const spotifyUrl =
      'https://accounts.spotify.com/authorize?' +
      new URLSearchParams({
        response_type: 'code',
        show_dialog: 'true',
        state: encodeURIComponent(state),
        client_id: encodeURIComponent(config.clientId),
        redirect_uri: redirectUri,
        scope: encodeURIComponent(config.scopes),
      }).toString();

    console.info('Please click the link to login to Spotify in the browser\n');
    console.info(spotifyUrl + '\n');

    const authUrl = await getLocalhostUrl(config.port);
    const params = new URLSearchParams(authUrl);
    const receivedCode = params.get('code');
    const receivedState = params.get('state');

    if (receivedState !== state) {
      goodBye('Received and original state do not match');
    }

    if (!receivedCode) {
      goodBye('No code received');
    }

    console.info('Login successfull!');

    const tokenRequestBody = new URLSearchParams({
      grant_type: 'authorization_code',
      code: receivedCode,
      redirect_uri: redirectUri,
    });

    const token: SpotifyTokenResponse = await request(
      {
        hostname: 'accounts.spotify.com',
        path: '/api/token',
        method: 'POST',
        headers: {
          'Content-type': 'application/x-www-form-urlencoded',
          Authorization:
            'Basic ' +
            Buffer.from(config.clientId + ':' + config.clientSecret).toString(
              'base64'
            ),
        },
      },
      tokenRequestBody.toString()
    );

    token.date_obtained = new Date().toUTCString();
    const outDir = write(config.outDir, config.outFileName, token);
    console.info("Success! Saved Spotify access token to '%s'", outDir);
  } catch (e) {
    goodBye('Something went wrong: ' + e);
  }
}
