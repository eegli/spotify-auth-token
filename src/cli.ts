import { authorize } from './authorize';
import { exit } from './utils';

export const cli = async () => {
  if (process.argv.length < 3) {
    exit(
      `
spotify-auth-token

- A command line utility for generating Spotify refresh tokens.

For docs & help, visit https://github.com/eegli/spotify-auth-token.
    `,
      'yellow'
    );
  }

  await authorize();
};
