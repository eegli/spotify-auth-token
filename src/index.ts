import { authorize } from './authorize';

export type { UserConfig } from './types';
export default authorize;
