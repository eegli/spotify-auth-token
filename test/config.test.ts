import { configFactory, defaultConfig } from '../src/config';

jest.spyOn(global.console, 'error').mockImplementation(jest.fn());

const createConfig = configFactory(defaultConfig, 'clientId', 'clientSecret');

describe('createConfig, programmatic use', () => {
  [
    { clientId: 'cid', clientSecret: 'cs' },
    {
      clientId: 'cid',
      clientSecret: 'cs',
      port: 1000,
      scopes: 'scopes',
      outDir: 'secrets',
      outFileName: 'mytoken',
    },
  ].forEach((config, idx) => {
    it(`works #${idx}`, () => {
      expect(createConfig(config)).toMatchSnapshot();
    });
  });

  [
    { clientSecret: 'cs' },
    {
      clientId: 'cid',
    },
  ].forEach((config, idx) => {
    it(`fails #${idx}`, () => {
      expect(() => {
        createConfig(config);
      }).toThrow();
    });
  });
});

describe('createConfig, cli use', () => {
  [
    ['--clientId', '111', '--clientSecret', '111'],
    ['--clientId', '111', '--clientSecret', '111', '--show'],
    ['--clientId', '333', '--clientSecret', '333', '--show', '--port', '4000'],
    ['--clientId', '777', '--clientSecret', '888', '--random', 'nothing'],
    [
      '--clientId',
      '222',
      '--clientSecret',
      '222',
      '--port',
      '1000',
      '--scopes',
      'scopes',
      '--outDir',
      'secrets/spotify',
      '--outFileName',
      'mytoken',
    ],
  ].forEach((config, idx) => {
    it(`works #${idx}`, () => {
      expect(createConfig(config)).toMatchSnapshot();
    });
  });
  [
    ['--clientId', 'cid'],
    ['--clientSecret', 'cs'],
  ].forEach((config, idx) => {
    it(`fails #${idx}`, () => {
      expect(() => {
        createConfig(config);
      }).toThrow();
    });
  });
});
