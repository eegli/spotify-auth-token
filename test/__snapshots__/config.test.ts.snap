// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createConfig, cli use works #0 1`] = `
Object {
  "clientId": "111",
  "clientSecret": "111",
  "outDir": "",
  "outFileName": "spotify-token",
  "port": 3000,
  "scopes": "user-read-email",
  "show": false,
}
`;

exports[`createConfig, cli use works #1 1`] = `
Object {
  "clientId": "111",
  "clientSecret": "111",
  "outDir": "",
  "outFileName": "spotify-token",
  "port": 3000,
  "scopes": "user-read-email",
  "show": true,
}
`;

exports[`createConfig, cli use works #2 1`] = `
Object {
  "clientId": "333",
  "clientSecret": "333",
  "outDir": "",
  "outFileName": "spotify-token",
  "port": 4000,
  "scopes": "user-read-email",
  "show": true,
}
`;

exports[`createConfig, cli use works #3 1`] = `
Object {
  "clientId": "777",
  "clientSecret": "888",
  "outDir": "",
  "outFileName": "spotify-token",
  "port": 3000,
  "scopes": "user-read-email",
  "show": false,
}
`;

exports[`createConfig, cli use works #4 1`] = `
Object {
  "clientId": "222",
  "clientSecret": "222",
  "outDir": "secrets/spotify",
  "outFileName": "mytoken",
  "port": 1000,
  "scopes": "scopes",
  "show": false,
}
`;

exports[`createConfig, programmatic use works #0 1`] = `
Object {
  "clientId": "cid",
  "clientSecret": "cs",
  "outDir": "",
  "outFileName": "spotify-token",
  "port": 3000,
  "scopes": "user-read-email",
  "show": false,
}
`;

exports[`createConfig, programmatic use works #1 1`] = `
Object {
  "clientId": "cid",
  "clientSecret": "cs",
  "outDir": "secrets",
  "outFileName": "mytoken",
  "port": 1000,
  "scopes": "scopes",
  "show": false,
}
`;
